### ISSUE TITLE
Redundant solid material evaluations

### ISSUE THREAD

**Issue by dragos-ana** [4C-multiphysics/4C#121]
## Description

The internal force and nonlinear stiffness computations for solid elements ( actions `Core::Elements::struct_calc_internalforce` and `Core::Elements::struct_calc_nlnstiff` in `evaluate` within `4C_solid_3D_ele_evaluate.cpp` ) use two separate calls to the solid material evaluation method `Mat::So3Material::evaluate`. This method computes both the stress and the linearization (and then disregards the one or the other depending on what we currently want to compute, see above). Therefore, we currently have a redundant material evaluation call. For materials which are computationally expensive, this results in significantly higher computation times than actually required. For example, saving the computed state and simply returning the saved linearization in the "redundant step" (while computing the stiffness contribution) led to a performance speedup of roughly factor 2 in my simulations with InelasticDefgradTransvIsotropElastViscoplast. Resolving this redundancy could therefore improve the performance significantly for expensive material evaluations.

## Possible Solution and Definition of Done

Based on a discussion with @amgebauer, we propose two possible solutions:
- [ ] Either fix this within all materials, by separating the stress and stiffness computations (e.g. replace `evaluate` by `evaluate_stress` and `evaluate_nlnstiff`, and maybe store them to not require recomputation for complex materials);
- [ ] Or store the computed contributions in the element evaluation and don't reevaluate them if they are available. This would be faster to implement, but it would lead to computing the stiffness contributions once too much per time step (per element).

I would propose a meeting for all interested parties, so that we can discuss what would be the more suitable approach. Also, feel free to add other proposals to the list. Reply under this thread if you are interested, so that we can organize this meeting.

**Comment by mayrmt** [comment]
@dragos-ana @amgebauer, how would this topic relate to nonlinear solvers that do not always require both residual and linearization, e.g. during a line search in a Newton method?

**Comment by sebproell** [comment]
What I have been doing to evaluate materials is based on designs in deal.II: Use evaluation flags to compute whatever you require and later query it.

Example:

```c++
    void
    reinit_point(unsigned int            q,
                 const SymmetricTensor2& linear_strain,
                 VectorizedArrayType     temperature,
                 const TsiEvalFlags&     flags)
  {
     if (contains(flags, TsiEvalFlags::stress))
       ...
     if (contains(flags, TsiEvalFlags::tanget_stress_strain))
       ...
     if (contains(flags, TsiEvalFlags::tanget_stress_strain))
       ...
```

where TsiEvalFlags are bit flags

```c++
enum class TsiEvalFlags
  {
    none                    = 0,
    update_history_in_place = 1 << 0,
    stress                  = 1 << 1,
    tangent_stress_strain   = 1 << 2,
  };

  // this generates operators like operator|(TsiEvalFlags,TsiEvalFlags)
  MacroAM_DefineBinaryOperators(TsiEvalFlags);
```

The computed values are stored and later queried by functions like

```c++
    const SymmetricTensor4&
    get_tangent_stress_strain() const
    {
      return stiffness;
    }
````

I found this approach to be very flexible so far. If I want to evaluate the residual, I pass only the `stress` flag. If I want to evaluate the tangent and residual at once, I pass `TsiEvalFlags::stress | TsiEvalFlags::tangent_stress_strain`. If I additionally know, that I want to update internal history based on the current state, I pass `update_history_in_place` as well.

**Comment by dragos-ana** [comment]
@mayrmt:
> @dragos-ana @amgebauer, how would this topic relate to nonlinear solvers that do not always require both residual and linearization, e.g. during a line search in a Newton method?
Due to the fact that both the contributions to residual and linearization are evaluated during the `Mat::So3Material::evaluate` call, this would still theoretically compute both of them (even though the linearization contributions would not be used). I would say that splitting this `Mat::So3Material::evaluate`  method into separate methods for residual and linearization would certainly speed up the computation also in the case of line search.


@sebproell 
This is certainly a nice approach and also similar to what we currently have with the actions within the solid evaluation. The main problem however is that there is one single material method which evaluates both contributions at the same time, so even though we want to compute the one or the other, we still force the material to compute both, pick the value we need (either stress or stiffness), and basically ignore the other. Maybe passing the flag to the material method would be a good idea though, to split the computation, as opposed to creating two separate methods (as I described above).

**Comment by sebproell** [comment]
@dragos-ana Maybe this wasn't clear enough: I pass this to the material ;) `reinit_point` is the equivalent of 4C's `Material::evaluate`.

**Comment by dragos-ana** [comment]
@sebproell 
> @dragos-ana Maybe this wasn't clear enough: I pass this to the material ;) `reinit_point` is the equivalent of 4C's `Material::evaluate`.

Got it now, thanks ;)

**Comment by mayrmt** [comment]
@dragos-ana That totally makes sense to me, however will also affect solid elements, since they have to pass this information from the model evaluator to the material.

**Comment by dragos-ana** [comment]
@mayrmt If I understand you correctly, you mean that the flag has to be passed consistently in the chain `model evaluator -> solid element -> material`, right?

**Comment by mayrmt** [comment]
@dragos-ana yes!

**Comment by amgebauer** [comment]
I updated the priority level to priority medium since solid mechanics tests seem to get measurable slower compared to the old framework.

**Comment by dragos-ana** [comment]
## Who is interested in discussing possible solutions?

As discussed in the Community Meeting on 25.02.2026, we want to first assemble developers / representatives for problem types directly affected by this issue. 
Please react to this comment with 🚀 if your problem type is affected and you want to discuss possible solutions to address it.

Afterwards, we will schedule a meeting together for the discussion. Thank you!

**Comment by amgebauer** [comment]
@4C-multiphysics/developers A short reminder to react with 🚀  to the message above if you think you can contribute to a discussion about the problem that we discussed in the last community meeting. It would be good if we have a group representing all important solid modules for the initial meeting to not miss any aspects.

**Comment by isteinbrecher** [comment]
Does this issue also affect out beam functionality / beam interaction?

**Comment by amgebauer** [comment]
@isteinbrecher Yes, any solution affects all modules that live within the new structural framework since it is about the core how we evaluate the residual and the linearization.

**Comment by isteinbrecher** [comment]
Then I don't understand the title and description of this issue, as it indicates that this only affects solid material evaluations. For beams we don't use the `Mat::So3Material::evaluate`.

**Comment by amgebauer** [comment]
yes, the title and the description might seem that the issue only affects solid element evaluation. Let me summarize the actual problem that I see here:

For the nonlinear solver, we need the residuum and (often) the linearization. In the old time integration framework, we always evaluated both, which is fine with a simple Newton algorithm. Only for the last Newton iteration, we unnecessarily evaluated the linearization. In the new framework, we evaluate the residuum, then we check for convergence, and if not yet converged, we evaluate the linearization to determine the the next search direction. However, the framework is not prepared for this, it does a complete reevaluation the linearization (and often the residuum). This might not be harmful in many situations, since evaluating and assembling the residuum is often a lot cheaper than its linearization and is often negligible with the other cost to solve the linear system. However, we currently do not rely on the fact that we just want to additionally want to have the linearization given that we already evaluated the residuum (accross all modules in the framework). We have material models (Dragos, Me, Paul) or element formulations (e.g., EAS) that are expensive to evaluate and we are currently evaluating them twice for each Newton iteration step doubling the evaluation cost. And these are just the cases, that we have previously been detected. Probably more are affected by this.

I see two possible workarounds:

 * Guarantee by design that when evaluating the linearization, we can optimize for the fact that we already evaluated the residuum. It is guaranteed by nox, but not within our the solid framework, where we do the whole evaluation again.
 * Optionally, always evaluate residuum and linearization in combination. It involves extra cost in cases with many line-search iterations. However, with our current way to evaluate the structure framework, it should still be better in many situations.

**Comment by dragos-ana** [comment]
Another summary of the issue, as well as the proposed solutions explained by @amgebauer, can be found in the [presentation](https://github.com/4C-multiphysics/community-meeting/issues/18#issuecomment-3960381445) I added to the Community Meeting Issue on 25.02.2026. The title is misleading because it is the original title of the issue I discovered at that time --- later it became clear that this is actually a problem of the new structural time integration itself which is quite significant in regards to computational performance.

**Comment by isteinbrecher** [comment]
Thanks for the explanation @amgebauer 

Is it possible to configure NOX such that it behaves like the old time integration, i.e., always evaluate force and stiff and reuse the stiffness if needed in a subsequent step?

**Comment by amgebauer** [comment]
I don't know whether NOX directly implements this out of the box. If not, we could also implement this on our own

**Comment by mayrmt** [comment]
As far as I know, NOX calls two separate functions `computeF` and `computeJacobian`. From a NOX point of view, that makes sense, since it provides a variety of nonlinear solvers, which not all need Jacobians (e.g. BFGS or lines search).

In monolithic FSI, we evaluate both during `computeF` and store the Jacobian, so that we can just return it (without evaluation) in `computeJacobian`.

**Comment by dragos-ana** [comment]
### Meeting Proposal: ~~18.03.2026, 3:00 PM~~ --> 24.03.2026, ~~4:30 PM~~ 2:00 PM

Thanks for your input on the poll above and on the issue in general. 
Based on the poll, I would propose a meeting **Tuesday, 24.03.2026, at ~~4:30 PM~~ 2:00 PM.**  
Please write me a private message (either slack or email) if that time or day does not suit you; I will try to make this suitable for everyone involved.

**Comment by isteinbrecher** [comment]
I would like to attend but I am on conference next week. So I will be available from the 23rd on.

**Comment by mairehenke** [comment]
It’s the same for me. Can we maybe postpone the meeting by a week?

**Comment by dragos-ana** [comment]
Based on your feedback, I added an alternative proposal. As before, let me know if that is still not feasible.

**Comment by dragos-ana** [comment]
Due to an urgent meeting, I must either advance the meeting here to an earlier time, or postpone it. 
My proposal would be to advance it to **2:00 PM** on the same day **(Tuesday, 24.03.2026)**. 
Please let me know whether that works for everyone, and sorry for the inconvenience!

**Comment by c-p-schmidt** [comment]
> Due to an urgent meeting, I must either advance the meeting here to an earlier time, or postpone it. My proposal would be to advance it to **2:00 PM** on the same day **(Tuesday, 24.03.2026)**. Please let me know whether that works for everyone, and sorry for the inconvenience!

For me it would also work at 2 pm
